### PR TITLE
collection of small fixes, features, and random pieces of code

### PR DESCRIPTION
**Pull request by hay-k**
This PR brings the following:

- some cherry picked commits from branch **alvaro/latest_20231215** that collectively
  1. introduce support for couplers in the qblox driver.
  2. expose `offset_i` and `offset_q` mixer calibration parameters for qblox RF modules.
- cleanup of the said commits to 
  1. align the code with main (the commits were based on older version of qibolab)
  2. make the code more pythonic and more readable
  3. remove things that were done from a perspective of a user that uses qblox alone (i.e. from a perspective where qibolab==qblox driver), but should not be part of things qibolab does.
- fix "deserialization" of `Custom` pulse shape:
  - as a side effect fixes deserialization of IIR, and SNR as well.

This is intended to cover the discussions in https://github.com/qiboteam/qibolab/issues/919 and https://github.com/qiboteam/qibolab/issues/921

### CODE REVIEW

**Comment by DavidSarlle** [comment]
@hay-k I am still reviewing the PR. Sorry, but after the last warm up I had to re calibrate the chip using the old driver in order to compare the results with your version of the qblox driver. I am finishing the 2q gates so, I will report what I see at the end of the week.

**Comment by DavidSarlle** [comment]
Hi @hay-k I am trying to reproduce the same results that we are getting with the old version for the driver and your PR. 

I am trying to reproduce the same results for the 2q gates running the iSWAP and CZ chevron. This routines are working perfectly with the old version of the driver:

## Classify q1 q2
---------------------------

  - id: single shot classification
    priority: 0
    next: chevron signal 
    operation: single_shot_classification
    targets: [1,2]
    parameters:
     nshots: 5_000
     relaxation_time: 100_000

  - id: chevron signal iSWAP
    operation: chevron_signal
    targets: [[1, 2]]
    parameters:
      amplitude_max_factor: 1.145 #0.23
      amplitude_min_factor: 0.9 #0.18
      amplitude_step_factor: 0.001
      duration_max: 180
      duration_min: 1
      duration_step: 1
      native: iSWAP
      nshots: 100
      relaxation_time: 100000
------------------------------------------------
      
With the latest version of the runcard (same used in the old driver):
https://github.com/qiboteam/qibolab_platforms_qrc/pull/161

And it is throwing the error:

Traceback (most recent call last):
  File "/nfs/users/david.fuentes/main/bin/qq", line 8, in <module>
    sys.exit(command())
  File "/nfs/users/david.fuentes/main/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/nfs/users/david.fuentes/main/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/nfs/users/david.fuentes/main/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/nfs/users/david.fuentes/main/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nfs/users/david.fuentes/main/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/nfs/users/david.fuentes/qibocal/src/qibocal/cli/_base.py", line 75, in auto
    autocalibrate(runcard, folder, force, update)
  File "/nfs/users/david.fuentes/qibocal/src/qibocal/cli/autocalibration.py", line 42, in autocalibrate
    history = runcard.run(
  File "/nfs/users/david.fuentes/qibocal/src/qibocal/auto/runcard.py", line 63, in run
    instance.run_protocol(
  File "/nfs/users/david.fuentes/qibocal/src/qibocal/auto/execute.py", line 131, in run_protocol
    completed = task.run(platform=self.platform, targets=self.targets, mode=mode)
  File "/nfs/users/david.fuentes/qibocal/src/qibocal/auto/task.py", line 149, in run
    completed.data, completed.data_time = operation.acquisition(
  File "/nfs/users/david.fuentes/qibocal/src/qibocal/auto/operation.py", line 42, in wrapper
    out = func(*args, **kwds)
  File "/nfs/users/david.fuentes/qibocal/src/qibocal/protocols/two_qubit_interaction/chevron/chevron_signal.py", line 123, in _aquisition
    results = platform.sweep(
  File "/nfs/users/david.fuentes/qibolab/src/qibolab/platform/platform.py", line 308, in sweep
    new_result = instrument.sweep(
  File "/nfs/users/david.fuentes/qibolab/src/qibolab/instruments/qblox/controller.py", line 300, in sweep
    self._sweep_recursion(
  File "/nfs/users/david.fuentes/qibolab/src/qibolab/instruments/qblox/controller.py", line 503, in _sweep_recursion
    result = self._execute_pulse_sequence(
  File "/nfs/users/david.fuentes/qibolab/src/qibolab/instruments/qblox/controller.py", line 184, in _execute_pulse_sequence
    module.process_pulse_sequence(
  File "/nfs/users/david.fuentes/qibolab/src/qibolab/instruments/qblox/cluster_qcm_bb.py", line 394, in process_pulse_sequence
    sequencer.waveforms_buffer.add_waveforms(
  File "/nfs/users/david.fuentes/qibolab/src/qibolab/instruments/qblox/sequencer.py", line 94, in add_waveforms
    pulse.idx_range = self.bake_pulse_waveforms(
  File "/nfs/users/david.fuentes/qibolab/src/qibolab/instruments/qblox/sequencer.py", line 140, in bake_pulse_waveforms
    waveform = pulse_copy.envelope_waveform_i(SAMPLING_RATE)
  File "/nfs/users/david.fuentes/qibolab/src/qibolab/pulses.py", line 868, in envelope_waveform_i
    return self.shape.envelope_waveform_i(sampling_rate)
  File "/nfs/users/david.fuentes/qibolab/src/qibolab/pulses.py", line 752, in envelope_waveform_i
    raise ValueError(
ValueError: Length of envelope_i must be equal to pulse duration in samples

I have tried setting the qf and coupler pulse to 200ns, that is the same number of samples of the custom shape used during this routine, but same error is raised.


It would be nice to check this routine, because it involves many of the fixes that you have introduced. 

Any idea what is the problem?

**Comment by hay-k** [comment]
@DavidSarlle is this experiment sweeping the duration of the custom pulse?

**Comment by DavidSarlle** [comment]
> @DavidSarlle is this experiment sweeping the duration of the custom pulse?

Yes, it is. The flux pulse is using custom shape, needed to get good chevrons. I know that the error is raised because the number of samples is changing during the sweeping. But this is something necessary and was fixed long time ago in Alvaro's branch because qblox does not support predistorted pulses yet.

By the way, I will write today a detailed report with the differences observed between branches.

**Comment by hay-k** [comment]
How is this type of sweep supposed to be interpreted? How does one sweep the duration of a pulse that is specified sample by sample and has fixed duration? 

I have checked what Alvaro has done in his branch. It pretty much boils down to [this](https://github.com/qiboteam/qibolab/blob/91ca64a452566bf2026e31be4415f187eb3bdb8a/src/qibolab/pulses.py#L618) line. The variable `num_samples` is the number of samples in the pulse duration. While the custom envelope always has 180 samples in it, the `num_samples` changes its value during the sweep. As long as `num_samples` is less than 180, this line of code takes only the first `num_samples` samples of the custom pulse. Thus, in one sweep iteration the custom pulse is the first 60 samples of the pulse definition, in another iteration it is the first 70 samples, etc. I am not sure if this is the correct way of sweeping custom pulse duration.

**Comment by DavidSarlle** [comment]
All results has been made with iqm5q qubit q1/q4 trying to minimizze the tiome between execution of the same routine.
The platform config and action runcard can be obtained from the qibo server using the report URL.
PRs used:

**(Old driver)**

qibocal -> alvaro/latest_20231215
qibolab -> alvaro/latest_20231215
qibolab_platforms_qrc -> david/alvaro_latest

**New driver:**

qibocal -> main
qibolab -> qblox-ad-hox-features
qibolab_platforms_qrc -> david/iqm5q_CZs_iSWAPS_refined


classify
--------

alvaro: http://login.qrccluster.com:9000/W4Bb2lEmRmSm1_BPuZLAaA==
hayk: http://login.qrccluster.com:9000/t7SaAt8IRU6wYP5qgQmRYQ==

COMMENTS: All results matching



qubit spec flux
----------------

alvaro: http://login.qrccluster.com:9000/EYdEAkTgRk2PWPA72WjohQ==
hayk: http://login.qrccluster.com:9000/PRWf2yC9R9mcZI1Iu-KJIw==

COMMENTS: Phase results not matching



qubit spec
-----------

alvaro: http://login.qrccluster.com:9000/eMJF7vRlQwKpH5heYWckHw==
hayk: http://login.qrccluster.com:9000/4BkgKYPCT72PXwD9LieYig==

COMMENTS: Phase results not matching





rabi amplitude
--------------

alvaro: http://login.qrccluster.com:9000/_kpHtQ08ReyxxRc_VpjAKg==
hayk: http://login.qrccluster.com:9000/egSiv8AVRzy3riUylYl4YA==

COMMENTS: Phase results not matching



rabi length
-----------

alvaro: http://login.qrccluster.com:9000/hHC9WWizS8a4TxmUDlwmzA==
hayk: http://login.qrccluster.com:9000/iMDueMb_SRSXsrDXmc1JCw==

COMMENTS: Probably RO pulse not well allocated during sweep and pulse sequence duration changed, as we pointed out long time ago.


ramsey
------

alvaro: http://login.qrccluster.com:9000/txpGzRxwRzG2dDG1H9gxRA==
hayk: http://login.qrccluster.com:9000/eOySVLI5RGCyzUl_IsgDmQ==

COMMENTS: Signal results not matching. Probably RO pulse not well allocated during sweep and pulse sequence duration changed, as we pointed out long time ago.



T1
--

alvaro: http://login.qrccluster.com:9000/A3ebtszyTmKpRrC9-BEu-A==
hayk: http://login.qrccluster.com:9000/Cl-y4DeiTP6_MBRlWBGt9Q==

COMMENTS: Signal results not matching. Probably RO pulse not well allocated during sweep and pulse sequence duration changed, as we pointed out long time ago.



T2
--

alvaro: http://login.qrccluster.com:9000/I8j2GDt6SMWqGwNplVkz2Q==
hayk: http://login.qrccluster.com:9000/a5IiAhEAQvez_GEHmBJsbQ==

COMMENTS: Signal results not matching. Probably RO pulse not well allocated during sweep and pulse sequence duration changed, as we pointed out long time ago.



OTHER PROBLEMS:
- Custom pulse shape not working when sweeping duration and modify the duration/samples.

**Comment by DavidSarlle** [comment]
> How is this type of sweep supposed to be interpreted? How does one sweep the duration of a pulse that is specified sample by sample and has fixed duration?
> 
> I have checked what Alvaro has done in his branch. It pretty much boils down to [this](https://github.com/qiboteam/qibolab/blob/91ca64a452566bf2026e31be4415f187eb3bdb8a/src/qibolab/pulses.py#L618) line. The variable `num_samples` is the number of samples in the pulse duration. While `self.envelope_i` always has 180 samples in it, the `num_samples` changes its value during the sweep. As long as `num_samples` is less than 180, this line of code takes only the first `num_samples` samples of the custom pulse. Thus, in one sweep iteration the custom pulse is the first 60 samples of the pulse definition, in another iteration it is the first 70 samples, etc. I am not sure if this is the correct way of sweeping custom pulse duration.

@aorgazf can give you more details. But that was the idea that we had to fix the problem. It works. as soon as we implemented, we were able to characterize the couplers and 2q gates. Before, the flux pulses were far away of being perfect and we needed to pre-distort them. As soon as we did it, we started to see the chevrons. @hay-k if you have any othe better solutions, feel free to introduce it and test it.

**Comment by alecandido** [comment]
> Thus, in one sweep iteration the custom pulse is the first 60 samples of the pulse definition, in another iteration it is the first 70 samples, etc. I am not sure if this is the correct way of sweeping custom pulse duration.

It is not: you should either compute the samples from a function, or interpolate the existing ones.

SciPy has easy-to-use interpolators that you could use for that
https://docs.scipy.org/doc/scipy/tutorial/interpolate.html

**Comment by aorgazf** [comment]
Hi,
I understand that sweeping the duration of a pulse with an arbitrary waveform could be done in multiple ways (including the interpolation that @alecandido mentioned). But for the 2q gate characterization protocols, I needed to play the first **i** samples of its waveform on each **iteration**. Iterating from 1 to the max number of samples of the waveform.

**Comment by DavidSarlle** [comment]
another example of results mismatching:

Landscape
---------

alvaro: can not upload it with qq upload (key error start time. Attached scrinshot)
imagen


hayk: http://login.qrccluster.com:9000/ugKLctRESUaccZpzn_c70w==

COMMENTS: Signal and/or phase results not matching. Probably RO pulse not well allocated during sweep and pulse sequence duration changed, as we pointed out long time ago.

We need to be able to run the landscape to fine tuning the CZs

**Comment by hay-k** [comment]
@DavidSarlle 

> Phase results not matching

I believe this is not an issue in the driver, it is because of some difference in data processing between the two versions of software.

> Probably RO pulse not well allocated

This was thoroughly discussed in https://github.com/qiboteam/qibolab/issues/919 and in a few meetings with @alvaro. The short summary is that this cannot be consistently fixed in qibolab until qibolab 0.2 is released. Why? The current qibolab API is limited and does not allow user to specify whether they want to move RO when the start or duration of other pulses is changed, or they want to keep RO in the same place always. Drivers cannot be implemented so that they handle both cases correctly - they always do the same thing regardless of what user requests (e.g. ZI driver always pushes the RO to the end, and the qblox driver always plays RO in a fixed place). It was concluded that for now it is a good consensus to implement all new features (e.g. couplers, mixer calibration, fix deserialization of Custom pulse, etc.), but don't fix the RO pulse start time sweep problems.

**Comment by hay-k** [comment]
Ayways, 

The original idea behind this PR was to implement things nicely, in a way that makes sense in general and not in the context of specific experiments and platforms only. Thus the PR could be merged and a new version of qibolab released. Given the fact that so far not all discussions/conclusions have been kept by all parties, and my opinion that there is no point in further discussions attempting to do things the proper way, I propose the following: 
- Drop the idea that this branch should be merged into main, and released. I hope this is fine by you, @scarrazza 
- Shoehorn whatever piece of code makes the most needed things work for setups with qblox driver, at the cost of making less important things broken (e.g. proper sweep of custom pulse duration, or experiments with duration/start sweeps that need RO in fixed location).
- Keep this branch updated with any changes in main as long as required.
- 0.2 release should have proper solutions to all problems that are the topic of this PR to bi considered having feature parity with 0.1*.

This means:
- Shoehorn some workaround code into the driver so that experiments like T2 and Ramsey work.
- Shoehorn some workaround code into the driver so that experiments like Rabi length work.
- Change the custom pulse envelope waveform to cut when the pulse duration is shorter than number of samples in the shape, pretending that this is a custom pulse duration sweep.

**Comment by scarrazza** [comment]
@hay-k at this stage please go ahead with this proposal. Qibolab 0.1 should contain the required patches until 0.2 takes over.

**Comment by hay-k** [comment]
@DavidSarlle 

In line with the proposal above, I have now pushed two commits to this branch:

1. for experiments where the start of a non-RO pulse is swept, the RO pulse will be pushed as well, which means experiments like T2 and Ramsey should produce expected results.
2. added the functionality to cut custom pulses to their first samples when their duration is swept, as discussed above.

Only remaining thing is the case where the RO needs to be pushed when the duration of a drive pulse is swept (e.g. Rabi length experiment). Turns out I didn't have a ready workaround for this one, so I will try to figure out something a bit later, and push a new commit.

**Comment by DavidSarlle** [comment]
Thanks @hay-k . Please try to reproduce the reports uploaded to be sure that the fixes introduced are working. 

Also remember the problems with displaying/manipulating the phase data in many routines. Try to reproduce yourself  the landscape report and other reports like qubit spec where this issue can be observed.

**Comment by hay-k** [comment]
Now I have a workaround for the case of duration sweeper as well, and Rabi length experiment works fine:

image

@DavidSarlle I didn't try the chevron experiment, can you test that one?

> Also remember the problems with displaying/manipulating the phase data in many routines.

Is this really a problem?

**Comment by hay-k** [comment]
> Is this really a problem?

Nevermind. I found where this difference is coming from. It should not cause any problems because this phase calculations are done only for the plots in the signal experiments, but anyways I changed it so that it looks like as it did in the past.

**Comment by DavidSarlle** [comment]
Thanks @hay-k . I have tested right now th elandscape and it is still not working. Thanks for fixing the phase problem, but maybe there is still something else that i snot working as expected as you can see in the landscape

http://login.qrccluster.com:9000/FzsNOANnT3W6LeXULiOxew==

**Comment by DavidSarlle** [comment]
@hay-k I tested the chevrons, and now is not raising any bug, but the results are not good:

alvaro: http://login.qrccluster.com:9000/jyjXrMv6Q0qCFz8WbYKlKg==
hayk: http://login.qrccluster.com:9000/1JShL92OSn2uRgtahTywyQ==

Please note that the chevron routine is using all the fixes introduced.

**Comment by hay-k** [comment]
@DavidSarlle This is because your platform definition for the new version of software is incomplete - the platform does not know that it has couplers, which results into the coupler pulse not being applied. To fix this you need to apply the following change:

```diff
-        str(FOLDER), qubits, pairs, instruments, settings, resonator_type="2D"
+        str(FOLDER), qubits, pairs, instruments, settings, resonator_type="2D", couplers=couplers
```

With this change I can see nice chevron similar to what you get from the olde versions:
image

**Comment by DavidSarlle** [comment]
> @DavidSarlle This is because your platform definition for the new version of software is incomplete - the platform does not know that it has couplers, which results into the coupler pulse not being applied. To fix this you need to do the following change:
> 
> ```diff
>      return Platform(
> -        str(FOLDER), qubits, pairs, instruments, settings, resonator_type="2D"
> +        str(FOLDER), qubits, pairs, instruments, settings, resonator_type="2D", couplers=couplers
>      )
> ```
> 
> With this change I can see nice chevron similar to what you get from the old versions: image

Oook @hay-k ! Thanks. I did not notice the missing definition in the platform file. Thanks a lot. Now the results are the ones expected. 

By the way, did you try the landscape? Maybe the coupler definition also explains the differences. I will try asap.

**Comment by hay-k** [comment]
No, I forgot to test the landscape, but if you are already doing it I will wait for your results.

**Comment by DavidSarlle** [comment]
> No, I forgot to test the landscape, but if you are already doing it I will wait for your results.

@hay-k now the landscape is also working! 
http://login.qrccluster.com:9000/1uA-E_ZtSNWYECv0Rqli2g==

From my side, I think that everything has been fixed and we can merge the PR. As we have been told to use only main, and even if the fixes are going to be adapted to qibolab 0.2 release, we should merge this PR asap, so everybody can use and deploy things using main for qibolab/qibocal/qibolab_platforms_qrc (we must merge https://github.com/qiboteam/qibolab_platforms_qrc/pull/161 as soon as possible. It is pending for reviewers approval).

**Review by DavidSarlle — DISMISSED** [review]

**Comment by DavidSarlle** [comment]
> > No, I forgot to test the landscape, but if you are already doing it I will wait for your results.
> 
> @hay-k now the landscape is also working! http://login.qrccluster.com:9000/1uA-E_ZtSNWYECv0Rqli2g==
> 
> From my side, I think that everything has been fixed and we can merge the PR. As we have been told to use only main, and even if the fixes are going to be adapted to qibolab 0.2 release, we should merge this PR asap, so everybody can use and deploy things using main for qibolab/qibocal/qibolab_platforms_qrc (we must merge [qiboteam/qibolab_platforms_qrc#161](https://github.com/qiboteam/qibolab_platforms_qrc/pull/161) as soon as possible. It is pending for reviewers approval).

DO NOT 

> No, I forgot to test the landscape, but if you are already doing it I will wait for your results.

@hayk seems that some of the landscapes are raising different results in with your PR:

Landscape q2-q3:

alvaro: http://login.qrccluster.com:9000/sfZpIn1iTeqDU6NBEKtuaA==
hayk: http://login.qrccluster.com:9000/mtRDARJ0TZ--IDuMaIrNaA==

Can you check all landscapes? I will be on vacation from 25th of September to 04th of October, so I will not be able to test anything during the next week. The rest of the test that I have made seems to be working so far. So Landscape should be tested in depth before merging into main

**Comment by hay-k** [comment]
If the landscape is working for one pair and not for another, my first guess would be discrepancy in the used parameters. I will check based on your runcards. Have a nice vacation!

> So Landscape should be tested in depth before merging into main

This PR is never going to be merged to main, because of the reasons I mentioned in [this](https://github.com/qiboteam/qibolab/pull/993#issuecomment-2360760592) comment, so you will need to use this branch instead of main until main is fully functional 0.2.*

**Comment by DavidSarlle** [comment]
@hay-k I double checked the platform and the runcard for all the qubits and I could not find any difference. So, please, check a see if you find any difference. also talk with @aorgazf because the way of manipulating the pulses in the qblox base band instrument is different in the ports I and Q, and that can explain the difference. 

I did not test the chevron for q2-q3 and q2-q4. But please, also test them and try to reproduce the same results. 
Same for landscape. We need to be completely sure that everything works as expected.

Regarding the merge. The instructions directly from Ji are that we must use only main in all qibo packages (qibocal, qibolab and qibolab_platforms) in order to characterize. Also the rest of people/groups that want to use an stable chip, must use main. So, please discuss this with @aorgazf, andy and Fred. At this point, as far as I know, the changes must be merged into main to have something that works for everybody.

**Comment by hay-k** [comment]
The landscape should be fixed now! The problem was affecting couplers 3 and 4 only.

**Comment by DavidSarlle** [comment]
> If the landscape is working for one pair and not for another, my first guess would be discrepancy in the used parameters. I will check based on your runcards. Have a nice vacation!
> 
> > So Landscape should be tested in depth before merging into main
> 
> This PR is never going to be merged to main, because of the reasons I mentioned in [this](https://github.com/qiboteam/qibolab/pull/993#issuecomment-2360760592) comment, so you will need to use this branch instead of main until main is fully functional 0.2.*

@hayk, that is what I though, but I double checked the platform.py and the parameters.json and compared with the one used with the old driver and I could not find any difference. But maybe I missed  something. Thans for check it again while I am on vacactions.

Regarding merging into main. Please talk with @aorgazf, Andy and Fred because JI gave us the instructions that everybody must use only main to control stable chips.

**Comment by alecandido** [comment]
> Regarding merging into main. Please talk with @aorgazf, Andy and Fred because JI gave us the instructions that everybody must use only main to control stable chips.

As I explained in private, the directive you mention is meaningful, if interpreted as "use a stable and shared version, instead of resorting to private versions, deemed to get outdated".

However, that could be only the general message, at high-level. At the technical level, the message has to be conjugated in an appropriate shape. And this is even changing over time, according to the packaging and distribution stage (as it happens for every piece of software - maintenance is a non-trivial science, with a non-unique solution).

So, I'm telling you that "use main" now translates to "use releases". Still, this is only applicable when you don't need experimental features, before their stabilization.
This branch is yet experimental, though it is needed, so it should aim for stabilization. Still, we're in the mid of a transition (0.1 -> 0.2), and this limits the available person power, though at the same time the 0.1 line of development is almost halted.
We can consider merging this branch to https://github.com/qiboteam/qibolab/tree/0.1 (certainly not to https://github.com/qiboteam/qibolab/tree/main, which is 0.2), and we will eventually do that. Though @hay-k is the person who created it, and he's responsible to maintain it update with any https://github.com/qiboteam/qibolab/tree/0.1, so he's the best person to tell when it's time to merge, and then eventually release (in the 0.1 series).

If you need installation in production mode (for better reproducibility), or there is the need to promote the direct usage of this driver with third parties (people further detached from Qibo developers), we should certainly accelerate the merge of it. But this has to be balanced with the required care to prevent as much as possible releasing possibly breaking code.

In any case, the person responsible for managing the software development is @scarrazza, not directly JI. So, I would closely follow the agreed directives established in meetings with him (i.e. the Qibo meetings).

**Review by DavidSarlle — APPROVED** [review]

**Comment by alecandido** [comment]
@hay-k I'm not sure if you already encountered, but do not worry about `pre-commit`. There is a known issue, and for the time being we're just ignoring it (if nothing happens before, next week I'll take action...).

**Review by alecandido — COMMENTED** [review]
This PR is mostly affecting just Qblox, and that was confirmed to work by @DavidSarlle.

I checked the changes, and those beyond the scope of the driver itself are very limited. I'm reporting a sign in the phases below (to be discussed there), while `platform/platform.py` is affected only to tune flux pulses, in a backward compatible way.
`pulses.py` is more heavily changed, but mostly just for the (de)serialization and waveforms (which are only used by Qblox itself), as well as a method of the `PulseSequence`, which I also confirmed is only used by Qblox.

It should be ready to be merged, but I will wait just one more day for some final tests (just to be completely sure).

Thanks @hay-k for these fixes!
(btw, you didn't mark as ready for review, but I will leave you to merge in any case)

**Review by alecandido — APPROVED** [review]

**Comment by scarrazza** [comment]
@hay-k can we merge this today?

File: src/qibolab/result.py
Comment (hay-k) [reply src/qibolab/result.py:41]: @alecandido The reason is the numpy's definition of the `arctan2` function. The phase (or angle as in `np.angle`) of a complex number `a+jb` is `np.arctan2(b, a)` and not `np.arctan2(a, b)`.

**Pull request merged**
